src/test/regress/sql/multi_complex_expressions.sql

```diff
@@ -36,7 +36,71 @@ SELECT count(*) FROM lineitem
 SELECT count(*) FROM lineitem
 	WHERE (l_receiptdate::timestamp - l_shipdate::timestamp) > interval '5 days';
 
-SELECT count(*) FROM lineitem WHERE random() = 0.1;
+-- can push down queries where no columns present on the WHERE clause
+SELECT count(*) FROM lineitem WHERE random() = -0.1;
+
+-- boolean tests can be pushed down
+SELECT count(*) FROM lineitem
+	WHERE (l_partkey > 10000) is true;
+
+-- scalar array operator expressions can be pushed down
+SELECT count(*) FROM lineitem
+	WHERE l_partkey = ANY(ARRAY[19353, 19354, 19355]);
+
+-- some more scalar array operator expressions
+SELECT count(*) FROM lineitem
+	WHERE l_partkey = ALL(ARRAY[19353]);
+
+-- operator expressions involving arrays
+SELECT count(*) FROM lineitem
+	WHERE ARRAY[19353, 19354, 19355] @> ARRAY[l_partkey];
+
+-- coerced via io expressions can be pushed down
+SELECT count(*) FROM lineitem
+	WHERE (l_quantity/100)::int::bool::text::bool;
+
+-- case expressions can be pushed down
+SELECT count(*) FROM lineitem
+	WHERE (CASE WHEN l_orderkey > 4000 THEN l_partkey / 100 > 1 ELSE false END);
+
+-- coalesce expressions can be pushed down
+SELECT count(*) FROM lineitem
+	WHERE COALESCE((l_partkey/50000)::bool, false);
+
+-- nullif expressions can be pushed down
+SELECT count(*) FROM lineitem
+	WHERE NULLIF((l_partkey/50000)::bool, false);
+
+-- null test expressions can be pushed down
+SELECT count(*) FROM orders
+	WHERE o_comment IS NOT null;
+
+-- functions can be pushed down
+SELECT count(*) FROM lineitem
+	WHERE isfinite(l_shipdate);
+
+-- constatnt expressions can be pushed down
+SELECT count(*) FROM lineitem
+	WHERE 0 != 0;
+
+-- distinct expressions can be pushed down
+SELECT count(*) FROM lineitem
+	WHERE l_partkey IS DISTINCT FROM 50040;
+
+-- row compare expression can be pushed down
+SELECT count(*) FROM lineitem
+	WHERE row(l_partkey, 2, 3) > row(2000, 2, 3);
+
+-- combination of different expressions can be pushed down
+SELECT count(*) FROM lineitem
+	WHERE
+		  (l_quantity/100)::int::bool::text::bool AND
+		  CASE WHEN l_orderkey > 4000 THEN l_partkey / 100 > 1 ELSE false END AND
+		  COALESCE((l_partkey/50000)::bool, false) AND
+		  NULLIF((l_partkey/50000)::bool, false) AND
+		  isfinite(l_shipdate) AND
+		  l_partkey IS DISTINCT FROM 50040 AND
+		  row(l_partkey, 2, 3) > row(2000, 2, 3);
```

Comment:
> Hmm, for that we already have an open [bug](https://github.com/citusdata/citus/issues/374). Is that what you're asking me to add as a test?

Yes, pretty much. Seemed like I was right that that's a worthwhile to
test case ;)